src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,548 @@
+fn font_get_face(_caller: Caller<'_, ShapingData>, _font: u32) -> u32 {
+    // Nothing to do here. We do not use face tokens.
+    0
+}
+
+// fn font_get_glyph(font: u32, unicode: u32, uvs: u32) -> u32;
+// Returns the nominal glyph ID for the given codepoint, using the cmap table of the font to map Unicode codepoint (and variation selector) to glyph ID.
+fn font_get_glyph(caller: Caller<'_, ShapingData>, _font: u32, codepoint: u32, uvs: u32) -> u32 {
+    let Some(codepoint) = char::from_u32(codepoint) else {
+        return 0;
+    };
+
+    match (uvs, char::from_u32(uvs)) {
+        (0, _) | (_, None) => caller.data().font.glyph_index(codepoint),
+        (_, Some(uvs)) => caller.data().font.glyph_variation_index(codepoint, uvs),
+    }
+    .unwrap_or_default()
+    .0 as u32
+}
+
+// fn font_get_scale(font: u32, x_scale: *mut i32, y_scale: *mut i32);
+// Returns the scale of the current font.
+fn font_get_scale(mut caller: Caller<'_, ShapingData>, _font: u32, x_scale: u32, y_scale: u32) {
+    // Return upem as rustybuzz has no scale.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+    let upem = caller.data().font.units_per_em();
+
+    _ = memory.write(
+        &mut caller.as_context_mut(),
+        x_scale as usize,
+        &upem.to_le_bytes(),
+    );
+    _ = memory.write(
+        &mut caller.as_context_mut(),
+        y_scale as usize,
+        &upem.to_le_bytes(),
+    );
+}
+
+// fn font_get_glyph_extents(font: u32, glyph: u32, extents: *mut CGlyphExtents) -> bool;
+// Returns the glyph's extents for the given glyph ID at current scale and variation settings.
+fn font_get_glyph_extents(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    extents: u32,
+) -> u32 {
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut glyph_extents = hb_glyph_extents_t::default();
+    let ret = caller
+        .data()
+        .font
+        .glyph_extents(GlyphId(glyph as u16), &mut glyph_extents);
+    if ret {
+        _ = memory.write(
+            caller.as_context_mut(),
+            extents as usize,
+            bytemuck::bytes_of(&glyph_extents),
+        );
+    }
+
+    ret as u32
+}
+
+// fn font_glyph_to_string(font: u32, glyph: u32, str: *const u8, len: u32);
+// Copies the name of the given glyph, or, if no name is available, a string of the form gXXXX into the given string.
+fn font_glyph_to_string(
+    mut caller: Caller<'_, ShapingData>,
+    _font: u32,
+    glyph: u32,
+    str: u32,
+    len: u32,
+) {
+    // len is the assigned heap memory. We should not allocate more than that.
+    // Should not assume we are not writing over anything.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let mut name = caller
+        .data()
+        .font
+        .glyph_name(GlyphId(glyph as u16))
+        .map(ToOwned::to_owned)
+        .unwrap_or(format!("g{glyph:4}"));
```

Comment:
Does this produce `g0000`? Can you add a comment explaining this?

Reply:
I dont understand what the problem is in returning that? The names in the test fonts are used to search for things like `.fina` (for Arabic) to find glyphs with specific properties. Any glyph with name `gXXXX` would be equally useless.

Reply:
Are you answering to the right comment? All I was asking for is to document `format!("g{glyph:4}")`, because it's not immediately obvious what it does. Maybe write it as  `format!("g{:4}", glyph)`

Reply:
Oh I thought you meant specifically `g0000`. I will double check.